reference-implementation/lib/writable-stream.js

```diff
@@ -182,215 +190,242 @@ function WritableStreamAddWriteRequest(stream) {
-    readyPromiseIsPending = true;
-  }
+    WritableStreamRejectClosedPromiseIfAny(stream, stream._storedError);
 
-  if (wasAborted === true) {
-    stream._pendingAbortRequest._reject(reason);
-    stream._pendingAbortRequest = undefined;
+    return;
   }
 
-  if (state === 'errored') {
-    WritableStreamRejectClosedPromiseIfAny(stream);
+  assert(state === 'writable');
+
+  stream._state = 'errored';
+  stream._storedError = reason;
+
+  if (stream._pendingAbortRequest === undefined) {
+    const writer = stream._writer;
+    if (writer !== undefined) {
+      defaultWriterReadyPromiseResetToRejected(writer, reason);
+      writer._readyPromise.catch(() => {});
+    }
+
+    WritableStreamRejectClosedPromiseIfAny(stream, reason);
 
     return;
   }
 
-  assert(state === 'closing');
+  stream._pendingAbortRequest._reject(reason);
+  stream._pendingAbortRequest = undefined;
 
-  stream._state = 'errored';
-  stream._storedError = reason;
+  WritableStreamRejectClosedPromiseIfAny(stream, reason);
+}
+
+function WritableStreamCloseInQueueOrInFlight(stream) {
+  if (stream._closeRequest === undefined && stream._inFlightCloseRequest === undefined) {
+    return false;
+  }
 
-  if (wasAborted === false && stream._writer !== undefined) {
-    WritableStreamDefaultWriterEnsureReadyPromiseRejectedWith(stream._writer, reason, readyPromiseIsPending);
+  return true;
+}
+
+function WritableStreamHasOperationMarkedInFlight(stream) {
+  if (stream._inFlightWriteRequest === undefined && stream._inFlightCloseRequest === undefined) {
+    return false;
   }
 
-  WritableStreamRejectClosedPromiseIfAny(stream);
+  return true;
+}
+
+function WritableStreamMarkCloseRequestInFlight(stream) {
+  assert(stream._inFlightCloseRequest === undefined);
+  assert(stream._closeRequest !== undefined);
+  stream._inFlightCloseRequest = stream._closeRequest;
+  stream._closeRequest = undefined;
 }
 
-function WritableStreamMarkFirstWriteRequestPending(stream) {
-  assert(stream._pendingWriteRequest === undefined, 'there must be no pending write request');
+function WritableStreamMarkFirstWriteRequestInFlight(stream) {
+  assert(stream._inFlightWriteRequest === undefined, 'there must be no pending write request');
   assert(stream._writeRequests.length !== 0, 'writeRequests must not be empty');
-  stream._pendingWriteRequest = stream._writeRequests.shift();
+  stream._inFlightWriteRequest = stream._writeRequests.shift();
 }
 
-function WritableStreamRejectClosedPromiseIfAny(stream) {
+function WritableStreamRejectClosedPromiseIfAny(stream, reason) {
   const writer = stream._writer;
   if (writer !== undefined) {
-    defaultWriterClosedPromiseReject(writer, stream._storedError);
+    defaultWriterClosedPromiseReject(writer, reason);
     writer._closedPromise.catch(() => {});
   }
 }
 
-function WritableStreamRejectPromisesInReactionToError(stream) {
+function WritableStreamRejectPromisesInReactionToError(stream, reason) {
```

Comment:
I struggled to have all the code look at the right thing (variables before/after update) when fixing the order of promise rejection/resolution. Saving old variables to `oldBlah` would also help but mess the code too. So, I've just let all the helpers take the new value directly than depending on that the caller makes sure to update the corresponding variables (e.g. `stream._storedError = reason`) before calling the helpers.

One benefit of letting this helper depend on `[[storedError]]` helps readers understand the invariants, i.e. the pending requests get rejected with the same exception as one in `[[storedError]]`. I understand that. So, there's trade-off.